src/trace_segment.ml

```diff
@@ -0,0 +1,799 @@
+    | { parent = This parent; _ } ->
+      iter_until_rev parent stop_symbol ~f;
+      f t
+  ;;
+
+  module Sentinel = struct
+    type nonrec t = t
+
+    let sentinel_location : Location.t =
+      { instruction_pointer = 0L; symbol_offset = 0; symbol = Unknown }
+    ;;
+
+    let[@inline always] create () = { location = sentinel_location; parent = Null }
+
+    let become_frame t location ~parent =
+      t.location <- location;
+      t.parent <- This parent;
+      t
+    ;;
+  end
+
+  module For_testing = struct
+    let rec to_string_list acc t =
+      match t.parent with
+      | Null -> acc
+      | This parent ->
+        to_string_list (Symbol.display_name t.location.symbol :: acc) parent
+    ;;
+
+    let to_string_list t = to_string_list [] t
+    let print_callstack leaf = to_string_list leaf |> String.concat_lines |> print_endline
+  end
+end
+
+module Control_flow = struct
+  type t =
+    | Jump
+    | Call
+    | Return
+end
+
+module Callstack = struct
+  type t =
+    #{ time : Timestamp.t
+     ; leaf : Frame.t
+     ; control_flow : Control_flow.t
+     }
+end
+
+type t =
+  { mutable root : Frame.Sentinel.t
+  ; callstacks : Callstack.t Nonempty_vec.t
+  }
+
+let create () =
+  let root = Frame.Sentinel.create () in
+  { root
+  ; callstacks =
+      Nonempty_vec.create
+        (#{ time = Timestamp.zero; leaf = (root :> Frame.t); control_flow = Return }
+         : Callstack.t)
+  }
+;;
+
+let[@inline always] current_frame t = (Nonempty_vec.last t.callstacks).#leaf
+
+let replace_root t location =
+  let new_sentinel = Frame.Sentinel.create () in
+  let root =
+    Frame.Sentinel.become_frame t.root location ~parent:(new_sentinel :> Frame.t)
+  in
+  t.root <- new_sentinel;
+  root
+;;
+
+let handle_call (t : t) (time : Timestamp.t) ~(src : Location.t) ~(dst : Location.t) =
+  let control_flow : Control_flow.t = Call in
+  match Frame.find (current_frame t) src.symbol with
+  | This _ as src_frame ->
+    Nonempty_vec.push_back
+      t.callstacks
+      #{ time; leaf = Frame.create dst ~parent:src_frame; control_flow }
+  | Null ->
+    (* I would only expect this to happen if this call is the very first event in this
+       trace-segment. *)
+    let src_frame = replace_root t src in
+    Nonempty_vec.push_back
+      t.callstacks
+      #{ time; leaf = Frame.create dst ~parent:(This src_frame); control_flow }
+;;
+
+let handle_return (t : t) (time : Timestamp.t) ~src:_ ~(dst : Location.t) =
+  let control_flow : Control_flow.t = Return in
+  match Frame.find (current_frame t) dst.symbol with
+  | This { parent; _ } ->
+    Nonempty_vec.push_back
+      t.callstacks
+      #{ time; leaf = Frame.create dst ~parent; control_flow }
+  | Null ->
+    (* We have returned into something we never saw the call for. This can happen if there
+       is a sequence of calls like [fn1 -> fn2 -> fn3] and we started tracing during the
+       execution of [fn2]. *)
```

Comment:
Nit: this comment doesn't actually describe this situation. The simplest example when this is `Null` is when we started when the callstack was `fn1 -> fn2`, and see a return transitioning from `src:fn2` to `dst:fn2`.

Reply:
> see a return transitioning from `src:fn2` to `dst:fn2`

Good point, I hadn't considered recursive functions that _aren't_ tail-recursive.